akd/src/directory.rs

```diff
@@ -233,68 +244,155 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
 
         let current_azks = self.retrieve_current_azks().await?;
         let current_epoch = current_azks.get_latest_epoch();
+        let lookup_info = self
+            .get_lookup_info::<H>(uname.clone(), current_epoch)
+            .await?;
+        let lookup_proof = self
+            .lookup_with_info::<H>(uname, &current_azks, current_epoch, lookup_info)
+            .await;
+        lookup_proof
+    }
 
+    async fn lookup_with_info<H: Hasher>(
+        &self,
+        uname: AkdLabel,
+        current_azks: &Azks,
+        current_epoch: u64,
+        lookup_info: LookupInfo,
+    ) -> Result<LookupProof<H>, AkdError> {
+        let current_version = lookup_info.value_state.version;
+
+        let lookup_proof = LookupProof {
+            epoch: current_epoch,
+            plaintext_value: lookup_info.value_state.plaintext_val,
+            version: lookup_info.value_state.version,
+            exisitence_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, false, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            existence_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.existent_label, current_epoch)
+                .await?,
+            marker_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, false, lookup_info.marker_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            marker_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.marker_label, current_epoch)
+                .await?,
+            freshness_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, true, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            freshness_proof: current_azks
+                .get_non_membership_proof(
+                    &self.storage,
+                    lookup_info.non_existent_label,
+                    current_epoch,
+                )
+                .await?,
+        };
+
+        // self.storage.log_metrics(log::Level::Info).await;
+
+        Ok(lookup_proof)
+    }
+
+    // TODO(eoz): Call proof generations async
```

Comment:
Can we file an issue to follow and discuss this? What needs to be called async?

Reply:
#166. So I think we can await on the completion of all lookups rather than awaiting for individual lookups and proceeding in the loop. Assuming my understanding of Rust's async is correct.

Reply:
Ah so the only thing that's "async" in the underlying logic here is the access to data-storage. Which assuming you've pre-cached all of the nodes will simply be accesses to memory storage but it can be read in parallel. I'm not sure how much we could gain by doing them all in parallel, but it should be trivial to do. You simply don't `.await` the futures and collect them in a `Vec`. Then you can await the batch of them with the `futures` crate

i.e. something like

```rust
use futures::future;

// generate your vector of individual tasks by spawning them on green threads
// but you need to clone "self" in order to apply on different threads, which will
// shard the data-access pointer and cache pointers, etc to yield individual Directory<_>
// instances on each thread
let tasks = proofs_to_generate.into_iter().map(|proof| {
  let self_clone = self.clone();
  tokio::spawn(async move { self_clone.lookup(proof).await })
}

let outputs = future::try_join_all(tasks).await?;
```

Reply:
Got it! I can try and add it in the next PR along with #167.